### PR TITLE
fix: disambiguate layout vs index route fetcher submissions

### PR DESCRIPTION
**Pull request by jacob-ebey**
Closes: https://github.com/remix-run/remix/issues/1291

- [ ] Docs
- [x] Tests

### CODE REVIEW

File: packages/remix-react/transition.ts
Comment (mjackson) [inline packages/remix-react/transition.ts:612]:
```suggestion
    for (let param of url.searchParams.getAll("index")) {
      if (!param) return true;
    }

    return false;
```

**Review by mjackson — COMMENTED** [review]
I'm not super familiar with this code so I'll defer to @ryanflorence

File: packages/remix-react/transition.ts
Comment (ryanflorence) [reply packages/remix-react/transition.ts:612]: Okay, this is weird I know ... but folks may need to use `?index=123` for bookmark pagination (or really anything), so the way we can do this without clobbering the `index` search param namespace is *only a bare index param means the index route*.

```
✅ /foo?index
✅ /foo?index&index=123
✅ /foo?index=123&index
❌ /foo?index=123
```

This code does exactly that, but it's not super clear since it's relying on "falsey" empty string and the fact that the loop happened at all. I'd go with:


```js
for (let param of url.searchParams.getAll("index")) {
  if (param === "") return true;
}
```

Comment (ryanflorence) [inline packages/remix-react/transition.ts:612]:
```suggestion
  function isIndexRequestUrl(url: URL) {
    for (let param of url.searchParams.getAll("index")) {
      // only use bare `?index` params without a value
      // ✅ /foo?index
      // ✅ /foo?index&index=123
      // ✅ /foo?index=123&index
      // ❌ /foo?index=123
      if (param === "") return true;
    }
    
    return false;
```

Comment (ryanflorence) [reply packages/remix-react/transition.ts:612]: Seems worth calling out for "future us".

File: integration/fetcher-layout-test.ts
Comment (ryanflorence) [inline integration/fetcher-layout-test.ts:81]: Can we get a test that tests `fetcher.load("/path?index")` and `fetcher.submit("/path?index")` here too? Or did I miss it?

**Review by ryanflorence — CHANGES_REQUESTED** [review]
Would like tests for `path?index` and a bit clearer intent in the loop :)

File: integration/fetcher-layout-test.ts
Comment (jacob-ebey) [reply integration/fetcher-layout-test.ts:106]: Loaders are in another ticket I'll do after this merges. There are changes that need to be made there. I've added a test for submitting to index loaders though.

File: packages/remix-react/transition.ts
Comment (jacob-ebey) [reply packages/remix-react/transition.ts]: Good call.

**Pull request merged**
